src/stash/src/transaction.rs

```diff
@@ -0,0 +1,641 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::{
+    cmp,
+    collections::{BTreeMap, BTreeSet, HashMap},
+    marker::PhantomData,
+    sync::{Arc, Mutex},
+};
+
+use futures::{
+    future::{self, try_join, try_join3, try_join_all, BoxFuture},
+    TryFutureExt,
+};
+use mz_ore::collections::CollectionExt;
+use serde_json::Value;
+use timely::{progress::Antichain, PartialOrder};
+use tokio::sync::mpsc;
+use tokio_postgres::Client;
+
+use crate::{
+    consolidate_updates_kv, postgres::CountedStatements, AntichainFormatter, AppendBatch, Data,
+    Diff, Id, InternalStashError, Stash, StashCollection, StashError, Timestamp,
+};
+
+impl Stash {
+    pub async fn with_transaction<F, T>(&mut self, f: F) -> Result<T, StashError>
+    where
+        F: FnOnce(Transaction) -> BoxFuture<Result<T, StashError>> + Clone + Sync + Send + 'static,
+    {
+        let (res, mut cons_rx) = self
+            .transact(|stmts, client| {
+                let f = f.clone();
```

Comment:
`transact` requires a `Fn` because it can execute it multiple times if it needs to retry, and `f` is moved into the async block below. Cloning `f` allows the clone to be moved, and so we can execute it as many times as we need. I think the mental model is: we need a closure where we are able to Clone anything the closure closes over because it might get async executed any number of times. @guswynn can maybe comment if I got that wrong.